File: Problems/No.1382_Balance_a_Binary_Search_Tree/balance_a_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    // get inorder traversal of the BST
    void InOrderTraverse(TreeNode* node, std::vector<int> &inorder) {

        if (node == NULL) return;
        InOrderTraverse(node->left, inorder);
        inorder.push_back(node->val);
        InOrderTraverse(node->right, inorder);

    }

    // construct a new Balanced Binary Search Tree
    TreeNode *ConstructBBST(std::vector<int> &inorder, int left, int right) {

        if (left > right) return NULL;
        else {

            int mid = left + (right-left)/2;
            TreeNode *newNode = new TreeNode(inorder[mid]);
            newNode->left = ConstructBBST(inorder, left, mid-1);
            newNode->right = ConstructBBST(inorder, mid+1, right);

            return newNode;

        }

    }

    TreeNode* balanceBST(TreeNode* root) {

        // parameters
        std::vector<int> InOrder;
        TreeNode *BBST = NULL;

        // start computing
        InOrderTraverse(root, InOrder);
        BBST = ConstructBBST(InOrder, 0, InOrder.size()-1);

        return BBST;

    }
};

File: Problems/No.1382_Balance_a_Binary_Search_Tree/balance_a_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def balanceBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """

        # get inorder traversal of the BST
        def InOrderTraverse(node):
            if node:
                InOrderTraverse(node.left)
                InOrder.append(node.val)
                InOrderTraverse(node.right)
        
        # construct a new Balanced Binary Search Tree
        def ConstructBBST(left, right):
            if (left > right):
                return None
            else:
                mid = left + (right-left)//2
                newNode = TreeNode(InOrder[mid])
                newNode.left = ConstructBBST(left, mid-1)
                newNode.right = ConstructBBST(mid+1, right)
                return newNode
        

        # parameters
        InOrder = []
        BBST = None

        # start computing
        InOrderTraverse(root)
        BBST = ConstructBBST(0, len(InOrder)-1)

        return BBST
